File: 0604/solution.ts
function productExceptSelf(nums: number[]): number[] {
    let len = nums.length;
    let left = new Array(len);
    left[0] = 1;
    let output = new Array(len);
    output[len - 1] = 1;

    for (let i = 1; i < len; i++) {
        left[i] = left[i - 1] * nums[i - 1];
        output[len - 1 - i] = output[len - i] * nums[len - i]
    }

    for (let i = 0; i < len; i++) {
        output[i] = left[i] * output[i]
    }
    return output
}

File: 0604/solution.go
package main

import "fmt"

func main() {
	numbs := []int{1, 2, 3, 4}
	self := productExceptSelf(numbs)

	fmt.Println(self)
}

func productExceptSelf(nums []int) []int {
	length := len(nums)
	var output = make([]int, length)
	var left = make([]int, length)
	left[0] = 1
	output[length-1] = 1

	for i := 1; i < length; i++ {
		left[i] = left[i-1] * nums[i-1]
		output[length-1-i] = output[length-i] * nums[length-i]
	}

	for i := 0; i < length; i++ {
		output[i] = left[i] * output[i]
	}
	return output
}
